web-search-engines/langchain4j-web-search-engine-searchapi/src/main/java/dev/langchain4j/web/search/searchapi/SearchApiWebSearchEngine.java

```diff
@@ -0,0 +1,120 @@
+import dev.langchain4j.web.search.WebSearchRequest;
+import dev.langchain4j.web.search.WebSearchResults;
+import lombok.Builder;
+
+import java.net.URI;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static dev.langchain4j.internal.Utils.getOrDefault;
+import static dev.langchain4j.internal.ValidationUtils.ensureNotBlank;
+import static java.time.Duration.ofSeconds;
+
+/**
+ * An implementation of a {@link WebSearchEngine} that uses
+ * <a href="https://www.searchapi.io">Search API</a> for performing web searches. <p>
+ * Search API supports not only Google Search, but many other engines
+ */
+public class SearchApiWebSearchEngine implements WebSearchEngine {
+
+    private static final String BASE_URL = "https://www.searchapi.io";
+    private static final String DEFAULT_ENGINE = "google";
+
+    private final String apiKey;
+    private final String engine;
+    private final SearchApiClient client;
+    private final Map<String, Object> optionalParameters;
+
+    /**
+     * @param apiKey             the Search API key for accessing their API
+     * @param timeout            the timeout duration for API requests
+     *                           <p>
+     *                           Default value is 30 seconds.
+     * @param engine             the engine used by Search API to execute the search
+     *                           <p>
+     *                           Default engine is Google Search.
+     * @param optionalParameters optional parameters to be passed on every request of this the engine, they can be overridden by the WebSearchRequest additional parameters for matching keys
+     *                           <p>
+     *                           Check <a href="https://www.searchapi.io">Search API</a> for more information on available parameters for each engine
+     */
+    @Builder
+    public SearchApiWebSearchEngine(String apiKey,
+                                    Duration timeout,
+                                    String engine,
+                                    Map<String, Object> optionalParameters) {
+        this.apiKey = ensureNotBlank(apiKey, "apiKey");
+        this.engine = getOrDefault(engine, DEFAULT_ENGINE);
+        this.optionalParameters = getOrDefault(optionalParameters, new HashMap<>());
+        this.client = SearchApiClient.builder()
+                .baseUrl(BASE_URL)
+                .timeout(getOrDefault(timeout, ofSeconds(30)))
+                .build();
+    }
+
+    @Override
+    public WebSearchResults search(WebSearchRequest webSearchRequest) {
+        SearchApiWebSearchRequest request = SearchApiWebSearchRequest.builder()
+                .apiKey(apiKey)
+                .engine(engine)
+                .query(webSearchRequest.searchTerms())
+                .optionalParameters(optionalParameters)
+                .additionalRequestParameters(webSearchRequest.additionalParams())
```

Comment:
@czelabueno As far as I understood from @mark-searchapi's review it is not standard among all engines, that's why we chose to decouple any implementation detail for a specific engine. See: https://github.com/langchain4j/langchain4j/pull/1216#discussion_r1678307385

Reply:
@zambrinf actually not, `startPage` and others is part of [OpenSearch standard](https://github.com/dewitt/opensearch/blob/master/opensearch-1-1-draft-6.md#opensearch-description-elements) (for all engines) as it's documented in `WebSearchRequest` so you can take them from the request and parse them to the expected key-value for each engine (e.g. google:?page=1;, bing:?page=1, etc). Anyway these are optional params.
If you prefer delegate the responsibility to the user to put every params as `additionalParam`, user should to be aware about all params supported by SearchApi by engine. It'd work same way. My only observation is that code could help to the user to set common params in the `WebSearchRequest` to do it easier. Otherwise documentation must be clear and up to date.

Reply:
As I understood we will leave this way so I will resolve it, you can unresolve if necessary.